src/Data/Singletons/Internal.hs

```diff
@@ -97,6 +111,40 @@ class SingKind k where
 data SomeSing k where
   SomeSing :: Sing (a :: k) -> SomeSing k
 
+-- | A pattern synonym for going between a term and the corresponding
+--
+-- As a constructor: this takes a singleton to its demoted (base)
+-- type.
+--
+-- >>> :t AsSing @Bool
+-- AsSing @Bool :: Sing a -> Bool
+-- >>> AsSing SFalse
+-- False
+--
+-- As a pattern: It extracts a singleton from its demoted (base) type.
+--
+-- @
+-- singAnd :: Bool -> Bool -> SomeSing Bool
+-- singAnd bool1 bool2
+--   | AsSing singBool1 <- bool1
+--   , AsSing singBool2 <- bool2
+--   = SomeSing (singBool1 %:&& singBool2)
```

Comment:
Why not write this directly like so?

```haskell
singAnd (AsSing singBool1) (AsSing singBool2) = SomeSing (singBool1 %&& singBool2)
```

Also, note that I wrote `%&&`, not `%:&&`: we don't use colons for singled infix operators anymore (after #230).

Reply:
I found better uses of `withSomeSing` but it wasn't easy to minimize them into an example

Reply:
To be clear: I think the example you've given is fine. I'm just suggesting a more concise way to write it.

Reply:
I misread. Updating to your version